### PR TITLE
Aleks work on top standalone jenkinsfile ( CASSANDRA-18594 )

### PR DESCRIPTION
**Pull request by michaelsembwever**
CASSANDRA-18594

### CODE REVIEW

File: .build/build-checkstyle.xml [hunk 1]
```diff
@@ -41,7 +41,7 @@
                     failOnViolation="true">
             <formatter type="plain"/>
             <formatter type="xml" tofile="${checkstyle.report.file}"/>
-            <fileset dir="${build.src.java}" includes="**/*.java"/>
+            <fileset dir="${build.src.java}" includes="**/*.java" excludes="build/**/*.java" />
```
Comment (michaelsembwever) [inline .build/build-checkstyle.xml:44]: undo

File: .build/build-rat.xml [hunk 2]
```diff
@@ -91,6 +91,7 @@
                  <exclude NAME="debian/control"/>
                  <exclude NAME="debian/dirs"/>
                  <exclude NAME="debian/patches/series"/>
+                 <exclude NAME=".jenkins/Jenkinsfile"/>
```
Comment (michaelsembwever) [inline .build/build-rat.xml]: undo

File: .build/check-code.sh [hunk 3]
```diff
@@ -24,5 +24,5 @@ command -v ant >/dev/null 2>&1 || { echo >&2 "ant needs to be installed"; exit 1
 [ -f "${CASSANDRA_DIR}/build.xml" ] || { echo >&2 "${CASSANDRA_DIR}/build.xml must exist"; exit 1; }
 
 # execute
-ant -f "${CASSANDRA_DIR}/build.xml" check dependency-check
+ant -f "${CASSANDRA_DIR}/build.xml" check #dependency-check FIXME uncomment when working again
```
Comment (michaelsembwever) [inline .build/check-code.sh]: undo (I think this is working again…)

File: .build/docker/build-artifacts.sh [hunk 4]
```diff
@@ -18,5 +18,5 @@
 #
 # Creates the tarball artifact
 
-$(dirname "$0")/_docker_run.sh bullseye-build.docker build-artifacts.sh $1
+$(dirname "$0")/_docker_run.sh bullseye-build.docker build-artifacts.sh
```
Comment (michaelsembwever) [inline .build/docker/build-artifacts.sh]: undo

File: .build/docker/build-jars.sh
Comment (michaelsembwever) [inline .build/docker/build-jars.sh:1]: I think we can remove this file.

File: .build/docker/_docker_run.sh [hunk 5]
```diff
@@ -26,10 +26,14 @@
 #
 ################################
 
+set -x
```
Comment (michaelsembwever) [inline .build/docker/_docker_run.sh]: do we need debug anymore ?

File: .build/docker/run-tests.sh [hunk 6]
```diff
@@ -24,6 +24,8 @@
 #  Define env variable DISABLE_INNER_SPLITS to disable inner splitting.
 #
 
+set -x
```
Comment (michaelsembwever) [inline .build/docker/run-tests.sh]: do we need debug anymore ?

File: .build/docker/run-tests.sh [hunk 7]
```diff
@@ -112,19 +125,25 @@ if [[ ! -z ${JENKINS_URL+x} ]] && [[ ! -z ${NODE_NAME+x} ]] ; then
...
+        echo >&2 "Unsupported operating system, expected Linux or Darwin"
+        exit 1
+esac
 
 # figure out resource limits, scripts, and mounts for the test type
 case ${target} in
     # test-burn doesn't have enough tests in it to split beyond 8, and burn and long we want a bit more resources anyway
     "stress-test" | "fqltool-test" | "microbench" | "test-burn" | "long-test" | "cqlsh-test" )
-        [[ ${mem} -gt $((5 * 1024 * 1024 * 1024 * ${jenkins_executors})) ]] || { echo >&2 "tests require minimum docker memory 6g (per jenkins executor (${jenkins_executors})), found ${mem}"; exit 1; }
+        # [[ ${mem} -gt $((5 * 1024 * 1024 * 1024 * ${jenkins_executors})) ]] || { echo >&2 "tests require minimum docker memory 6g (per jenkins executor (${jenkins_executors})), found ${mem}"; exit 1; }
```
Comment (michaelsembwever) [inline .build/docker/run-tests.sh]: should this still be commented out ?

File: .build/docker/run-tests.sh [hunk 8]
```diff
@@ -178,7 +208,7 @@ TEST_SCRIPT=${test_script}
 JAVA_VERSION=${java_version}
 PYTHON_VERSION=${python_version}
 cython=${cython}
-ANT_OPTS="-Dtesttag.extra=.arch=$(arch).python${python_version}"
+ANT_OPTS="${ANT_OPTS}"
```
Comment (michaelsembwever) [inline .build/docker/run-tests.sh]: why ?

File: .build/docker/run-tests.sh [hunk 9]
```diff
@@ -213,8 +243,10 @@ docker exec --user root ${container_name} bash -c "\${CASSANDRA_DIR}/.build/dock
 docker exec --user root ${container_name} update-alternatives --set python /usr/bin/python${python_version} | tee -a ${logfile}
 
 # capture logs and pid for container
```
Comment (michaelsembwever) [inline .build/docker/run-tests.sh]: pid is not longer captured AFAIK
we are capturing status now, with pipefail

File: .jenkins/Jenkinsfile [hunk 10]
```diff
@@ -22,517 +22,447 @@
 // Validate/lint this file using the following command
 // `curl -X POST  -F "jenkinsfile=<.jenkins/Jenkinsfile" https://ci-cassandra.apache.org/pipeline-model-converter/validate`
 
+def jdksSupported = ["11", "17"] 
+def archsSupported = ["amd64", "arm64"]
+def pythonsSupported = ["3.7", "3.8", "3.11"] 
+def jdkDefault = "11"
+def pythonDefault = "3.8"
+def stageResults = [:]
+def splits = 10
```
Comment (michaelsembwever) [inline .jenkins/Jenkinsfile:30]: splits differ for each test type

File: .jenkins/Jenkinsfile [hunk 11]
```diff
@@ -22,517 +22,447 @@
 // Validate/lint this file using the following command
 // `curl -X POST  -F "jenkinsfile=<.jenkins/Jenkinsfile" https://ci-cassandra.apache.org/pipeline-model-converter/validate`
 
+def jdksSupported = ["11", "17"] 
+def archsSupported = ["amd64", "arm64"]
+def pythonsSupported = ["3.7", "3.8", "3.11"]
```
Comment (michaelsembwever) [inline .jenkins/Jenkinsfile]: remove 3.7

File: .jenkins/Jenkinsfile [hunk 12]
```diff
@@ -22,517 +22,447 @@
...
+  'dtest-large-novnode': [type: 'test', splits: splits, dtest: true],
+  'dtest-upgrade': [type: 'test', splits: splits, dtest: true],
+  'dtest-upgrade-novnode': [type: 'test', splits: splits, dtest: true],
+  'dtest-upgrade-large': [type: 'test', splits: splits, dtest: true],
+  'dtest-upgrade-novnode-large': [type: 'test', splits: splits, dtest: true],
+]
+
+// define matrix axes
+def Map matrix_axes = [
+    arch: archsSupported,
```
Comment (michaelsembwever) [inline .jenkins/Jenkinsfile:65]: just "arm" for now
(can be done instead in `isArchEnabled()`)

File: .jenkins/Jenkinsfile [hunk 13]
```diff
@@ -22,517 +22,447 @@
...
+  tests: [failFast: false],
+]
+for (def axis in _axes) {
+  def cell = axis
+  def name = getStepName(cell, stepsMap[cell.step])
+  tasks[cell.step == "jar" ? "jars" : "tests"][name] = { ->
+    "${stepsMap[cell.step].type}"(stepsMap[cell.step], cell) // build() or test()
+  }
+}
+
```
Comment (michaelsembwever) [inline .jenkins/Jenkinsfile:102]: how much of the above scripting can we move to below line 180 (alongside the other scripting support code) ?

File: .jenkins/Jenkinsfile [hunk 14]
```diff
@@ -22,517 +22,447 @@
...
 pipeline {
-  agent { label 'cassandra' }
+  agent none
+  options {
+    skipDefaultCheckout()
+    // retry(3) // Do we need it?
+  }
+  parameters {
+    string(name: 'repository', defaultValue: 'https://github.com/HadesArchitect/cassandra', description: 'Cassandra Repository')
+    string(name: 'branch', defaultValue: '*/aleks-cassius', description: 'Branch')
```
Comment (michaelsembwever) [inline .jenkins/Jenkinsfile]: add comment that eventually this needs to match the current repo and branch… 
(is there some way we can check and enforce this?)

File: .jenkins/Jenkinsfile [hunk 15]
```diff
@@ -22,517 +22,447 @@
...
+    booleanParam(name: 'stage_jvm-dtest-upgrade', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_dtest', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_dtest-novnode', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_dtest-offheap', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_dtest-large', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_dtest-large-novnode', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_dtest-upgrade', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_dtest-upgrade-novnode', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_dtest-upgrade-large', defaultValue: false) // return default to true
+    booleanParam(name: 'stage_dtest-upgrade-novnode-large', defaultValue: false) // return default to true
```
Comment (michaelsembwever) [inline .jenkins/Jenkinsfile]: given the stages (and test types) are (/will be) dynamic, this booleanParam approach won't survive.

can we use either a `choice` or `stringParam` (the string param being like a regex) ?

**Comment by michaelsembwever** [comment]
finished first review round.  looking good, just minor stuff…

File: .build/docker/_docker_run.sh (hunk 5 above)
Comment (HadesArchitect) [reply .build/docker/_docker_run.sh]: Actually it's pretty convenient. Maybe I do it configurable (enabled if a verbosity flag is raised)?

File: .build/docker/run-tests.sh (hunk 6 above)
Comment (HadesArchitect) [reply .build/docker/run-tests.sh]: Same as for `_docker_run`, pretty convenient. Add as an optional?

File: .build/docker/run-tests.sh (hunk 8 above)
Comment (HadesArchitect) [reply .build/docker/run-tests.sh]: Oh, indeed. Must be a left-over from an experiment. Fix in a moment.

File: .build/docker/run-tests.sh [hunk 16]
```diff
@@ -213,8 +243,10 @@ docker exec --user root ${container_name} bash -c "\${CASSANDRA_DIR}/.build/dock
 docker exec --user root ${container_name} update-alternatives --set python /usr/bin/python${python_version} | tee -a ${logfile}
```
Comment (HadesArchitect) [inline .build/docker/run-tests.sh:238]: Shouldn't it also be pipefail?

File: .build/build-checkstyle.xml (hunk 1 above)
Comment (HadesArchitect) [reply .build/build-checkstyle.xml:44]: As far as I remember, checkstyle check was badly failing there with tons of warnings. Do the generated files also have to be validated?

Comment (michaelsembwever) [reply .build/build-checkstyle.xml:44]: oh, you are right.  generated java files should be ignored.

File: .build/docker/_docker_run.sh (hunk 5 above)
Comment (michaelsembwever) [reply .build/docker/_docker_run.sh]: behind a flag, yes please.

File: .build/docker/run-tests.sh (hunk 6 above)
Comment (michaelsembwever) [reply .build/docker/run-tests.sh]: behind a flag, yes please.

File: .build/docker/run-tests.sh (hunk 16 above)
Comment (michaelsembwever) [reply .build/docker/run-tests.sh:238]: yes

File: .jenkins/Jenkinsfile (hunk 13 above)
Comment (michaelsembwever) [reply .jenkins/Jenkinsfile:100]: I will tackle this later. I have some ideas…

File: .jenkins/Jenkinsfile [hunk 17]
```diff
@@ -22,517 +22,447 @@
...
-                }
-            }
+
+  // sh "mkdir -p build"
+  // def build_dir = sh( returnStdout: true, script:"mktemp -d ${WORKSPACE}/build/build_docker_run.XXXXXXXXXX" ).trim()
+  // def build_dir_rel = build_dir - "${WORKSPACE}/"
+
+  // dir("${build_dir}") {
+    // fetchSource(cell.step, arch, jdk)
+  // }
```
Comment (michaelsembwever) [inline .jenkins/Jenkinsfile:263]: what's going on here ?

File: .build/docker/run-tests.sh (hunk 16 above)
Comment (HadesArchitect) [reply .build/docker/run-tests.sh:238]: In case of fail we just quit with exit status 1, right?

Comment (michaelsembwever) [reply .build/docker/run-tests.sh:238]: this should just exit yes.   it is not expected to ever fail.

File: .jenkins/Jenkinsfile (hunk 12 above)
Comment (HadesArchitect) [reply .jenkins/Jenkinsfile:65]: You meant `amd`, right? I'll remove it from the list `archsSupported` as it should be "the single source of truth"

File: .jenkins/Jenkinsfile (hunk 10 above)
Comment (HadesArchitect) [reply .jenkins/Jenkinsfile:30]: It's the "generic" split count, they can be specified explicitly per stage in the `stepsMap`

File: .jenkins/Jenkinsfile (hunk 15 above)
Comment (HadesArchitect) [reply .jenkins/Jenkinsfile]: I planned to generate booleanParams based on the `stepsMap` but it's hard with Jenkins, need an additional plugin or some dark magic. Let's stick to stringParam for now, I guess.

File: .jenkins/Jenkinsfile (hunk 17 above)
Comment (HadesArchitect) [reply .jenkins/Jenkinsfile:263]: Oooops, sorry. Will clean up.

File: .jenkins/Jenkinsfile (hunk 12 above)
Comment (michaelsembwever) [reply .jenkins/Jenkinsfile:65]: amd yes

Comment (HadesArchitect) [reply .jenkins/Jenkinsfile:65]: Done

File: .jenkins/Jenkinsfile (hunk 13 above)
Comment (HadesArchitect) [reply .jenkins/Jenkinsfile:102]: Well, nearly all of it. Should I?

Comment (HadesArchitect) [reply .jenkins/Jenkinsfile:102]: I'd keep stepsMap and matrix definition upstairs for sure, maybe excludes too... Just for visibility. But no pushing, can move it down.

File: .jenkins/Jenkinsfile [hunk 18]
```diff
@@ -250,227 +198,149 @@ def isCanonical() {
...
-        def jdk = j
-        builds_per_jdk[jdk] = { _buildAxis(build_script, toCopy, arch, jdk) }
+def build(command, cell) {
+  def build_script = ".build/docker/${command.script}"
+  node(getNodeArch(cell.arch)) {
+    withEnv(cell.collect { k, v -> "${k}=${v}" }) {
+      ws("workspace/${JOB_NAME}/${BUILD_NUMBER}/${cell.step}/${cell.arch}/jdk-${cell.jdk}") {
+          cleanWs(disableDeferredWipeout: true)
+          fetchSource(cell.step, cell.arch, cell.jdk)
+          def logfile = "${JOB_NAME}_${BUILD_NUMBER}_${cell.step}_jdk${cell.jdk}_${cell.arch}.log"
```
Comment (michaelsembwever) [inline .jenkins/Jenkinsfile:223]: logfile needs to be under build/

File: .jenkins/Jenkinsfile [hunk 19]
```diff
@@ -250,227 +198,149 @@ def isCanonical() {
...
+          catchError(buildResult: null, message: 'Build task failed', stageResult: 'FAILURE') {
+              sh label: 'RUNNING BUILD TASK...', script: "#!/usr/bin/bash \nset -o pipefail; build_dir='${WORKSPACE}/build' DEBUG=true ${build_script} ${cell.jdk} 2>&1 | tee ${logfile}"
+              if ("jar" == step) { // only stash the project built files. all dependency libraries are restored from the local maven repo using `ant resolver-dist-lib`
+                  stash name: "${cell.arch}_${cell.jdk}", useDefaultExcludes: false //, includes: '**/*.jar' //, includes: "*.jar,classes/**,test/classes/**,tools/**"
+              }
+              copyToNightlies("${command.toCopy}", "${STAGE_NAME}/jdk${cell.jdk}/${cell.arch}/")
+          }
+          sh "xz -f *.log"
+          archiveArtifacts artifacts: "**/*.log.xz", fingerprint: true
+          copyToNightlies("**/*.log.xz", "${cell.step}/jdk${cell.jdk}/${cell.arch}/")
```
Comment (michaelsembwever) [inline .jenkins/Jenkinsfile:234]: can go inside a `dir("build") { … }`

File: .jenkins/Jenkinsfile [hunk 20]
```diff
@@ -250,227 +198,149 @@ def isCanonical() {
...
+def test(command, cell) {
+  def splits = command.splits ? command.splits : 1 
+  def python = cell.python
+  def cython = cell.cython
+  node(getNodeArch(cell.arch)) {
+    withEnv(cell.collect { k, v -> "${k}=${v}" }) {
+      ws("workspace/${JOB_NAME}/${BUILD_NUMBER}/${cell.step}/${cell.arch}/jdk-${cell.jdk}/python-${cell.python}") {
+        cleanWs(disableDeferredWipeout: true)
+        fetchSource(cell.step, cell.arch, cell.jdk)
+        def logfile = "${JOB_NAME}_${BUILD_NUMBER}_${cell.step}_jdk${cell.jdk}_python_${cell.python}_${cell.cython}_${cell.arch}.log"
```
Comment (michaelsembwever) [inline .jenkins/Jenkinsfile:250]: goes under `build/`

File: .jenkins/Jenkinsfile [hunk 21]
```diff
@@ -250,227 +198,149 @@ def isCanonical() {
...
+          dir("build") {
+            // junit testResults: "test/output/nosetests.xml", allowEmptyResults: true
+            // junit testResults: "test/output/cqlshlib.xml", allowEmptyResults: true
+            //sh "ls -la test/output/"
+            junit testResults: "test/output/**/TEST*.xml", allowEmptyResults: true
+            sh "xz -f test/output/**/TEST*.xml || true ; xz -f test/output/cqlshlib.xml test/output/nosetests.xml || true"
+            // junit testResults: "test/output/**/TEST*.xml,test/output/cqlshlib.xml,test/output/nosetests.xml", testDataPublishers: [[$class: 'StabilityTestDataPublisher']]
+            // junit testResults: "build/test/output/**/TEST*.xml,build/test/output/cqlshlib.xml,build/test/output/nosetests.xml"//, testDataPublishers: [[$class: 'StabilityTestDataPublisher']]
+            // sh "xz -f build/test/output/**/TEST*.xml || true ; xz -f build/test/output/cqlshlib.xml build/test/output/nosetests.xml || true"
```
Comment (michaelsembwever) [inline .jenkins/Jenkinsfile:275]: needs cleanup.

remove ` allowEmptyResults: true`
needs `testDataPublishers: [[$class: 'StabilityTestDataPublisher']]`

File: .jenkins/Jenkinsfile [hunk 22]
```diff
@@ -250,227 +198,149 @@ def isCanonical() {
...
-        } else {
-          echo "Completed ${STAGE_NAME} '${build_script} ${jdk}' on ${NODE_NAME}"
+        if (stageError) {
+          stageResults[getStepName(cell, command)] = "FAILURE"
         }
+        sh "xz -f *.log"
+        archiveArtifacts artifacts: "*.log.xz,build/**/*.log.xz,build/test/logs/**,build/test/output/**/TEST*.xml.xz,build/test/output/cqlshlib.xml.xz,build/test/output/nosetests.xml.xz", fingerprint: true
+        // archiveArtifacts artifacts: "build/**/*.log.xz,build/test/logs/**,build/test/output/**/TEST*.xml.xz,build/test/output/cqlshlib.xml.xz,build/test/output/nosetests.xml.xz", fingerprint: true
+        copyToNightlies("*.log.xz,build/test/logs/**", "${cell.step}/${cell.arch}/jdk${cell.jdk}/python${cell.python}/cython_${cell.cython}/" + "split_${cell.split}_${splits}".replace("/", "_"))
+
```
Comment (michaelsembwever) [inline .jenkins/Jenkinsfile:286]: can go inside a dir("build") { … }

File: .jenkins/Jenkinsfile [hunk 23]
```diff
@@ -506,13 +376,10 @@ def saveAgentReport() {
...
-    checkout scm
-    sh "mkdir -p build"
+  node(getNodeArch('amd64')) {
     copyArtifacts filter: 'test/output/**/*.xml.xz', fingerprintArtifacts: true, projectName: env.JOB_NAME, selector: specific(env.BUILD_NUMBER), target: "build/"
     sh 'xz -d build/test/output/**/*.xml.xz || true ; xz -d test/output/cqlshlib.xml test/output/nosetests.xml || true'
-    sh "ANT_OPTS='-Xmx15g' ant generate-test-report"
+    sh ".build/docker/_docker_run.sh bullseye-build.docker generate-test-report.sh"
     dir('build/') {
       sh "xz -f TESTS-TestSuites.xml"
       archiveArtifacts artifacts: "TESTS-TestSuites.xml.xz,test/html/**/*", fingerprint: true
```
Comment (michaelsembwever) [inline .jenkins/Jenkinsfile:385]: we'll need to make a zip file that matches apple's ci results bundle here…

File: .jenkins/Jenkinsfile [hunk 24]
```diff
@@ -506,13 +376,10 @@ def saveAgentReport() {
 /////////////////////////////////////////
 
 def generateUnifiedTestReport() {
-  assert "Summary".equals(STAGE_NAME)
-  node('cassandra') {
-    checkout scm
-    sh "mkdir -p build"
+  node(getNodeArch('amd64')) {
     copyArtifacts filter: 'test/output/**/*.xml.xz', fingerprintArtifacts: true, projectName: env.JOB_NAME, selector: specific(env.BUILD_NUMBER), target: "build/"
```
Comment (michaelsembwever) [inline .jenkins/Jenkinsfile:380]: this is not working for me…

File: .build/docker/_docker_run.sh [hunk 25]
```diff
@@ -26,10 +26,14 @@
 #
 ################################
 
+[ $DEBUG ] && set -x
+
 # variables, with defaults
 [ "x${cassandra_dir}" != "x" ] || cassandra_dir="$(readlink -f $(dirname "$0")/../..)"
 [ "x${build_dir}" != "x" ] || build_dir="${cassandra_dir}/build"
+[ "x${m2_dir}" != "x" ] || m2_dir="${cassandra_dir}/build/m2"
```
Comment (michaelsembwever) [inline .build/docker/_docker_run.sh:34]: this is problematic.  

we want the `~/.m2` directory to be used, so to re-use the already downloaded .m2/repository 
(we don't want to download everything every time unnecessarily)

Comment (HadesArchitect) [reply .build/docker/_docker_run.sh:34]: I understand the rationale but the Home directory had some problems, like being unwritable on jenkins agent or something like that. Maybe stash m2 things? Is there any another possible approach?

File: .jenkins/Jenkinsfile [hunk 26]
```diff
@@ -250,227 +198,149 @@ def isCanonical() {
...
-          parallel(tests_per_python)
-        }
-      }
-      parallel(tests_per_jdk)
-    }
-  }
-  parallel(tests_per_arch)
-}
+// def testJob(test_type, splits=1) {
+//   def jdks = STAGE_NAME.contains("dtest-upgrade") || "simulator-dtest".equals(STAGE_NAME) ? ["${javaVersionDefault}"] : "${javaVersionsSupported}".split(/,/, -1)
```
Comment (HadesArchitect) [inline .jenkins/Jenkinsfile:308]: Could you please explain me what's going on here? It's not implemented in my version yet.

File: .jenkins/Jenkinsfile (hunk 24 above)
Comment (HadesArchitect) [reply .jenkins/Jenkinsfile:380]: Thanks, I'll check again

File: .jenkins/Jenkinsfile (hunk 23 above)
Comment (HadesArchitect) [reply .jenkins/Jenkinsfile:385]: Can we have a quick call to discuss it?

File: .jenkins/Jenkinsfile (hunk 26 above)
Comment (michaelsembwever) [reply .jenkins/Jenkinsfile:308]: I don't know what that commented out lines is.  just delete.

File: .build/docker/_docker_run.sh (hunk 25 above)
Comment (michaelsembwever) [reply .build/docker/_docker_run.sh:34]: the problem is we want to re-use the `.m2/repository/` between runs, not just between stages… 

e.g. the post-commit CI on ci-cassandra.a.o

File: .jenkins/Jenkinsfile (hunk 24 above)
Comment (michaelsembwever) [reply .jenkins/Jenkinsfile:380]: I have it working on my branch `mck/aleks-cassius`

File: .jenkins/Jenkinsfile (hunk 23 above)
Comment (michaelsembwever) [reply .jenkins/Jenkinsfile:385]: let me chase this up with @dcapwell 

the generation of the correct zipfile might be all part of the generate-test-report.sh script

File: .jenkins/Jenkinsfile [hunk 27]
```diff
@@ -250,227 +198,149 @@ def isCanonical() {
...
+  // sh "mkdir -p build"
+  // def build_dir = sh( returnStdout: true, script:"mktemp -d ${WORKSPACE}/build/build_docker_run.XXXXXXXXXX" ).trim()
+  // def build_dir_rel = build_dir - "${WORKSPACE}/"
+
+  // dir("${build_dir}") {
+    // fetchSource(cell.step, arch, jdk)
+  // }
+        def stageError = true
+        catchError(buildResult: null, message: 'Tests failed', stageResult: 'FAILURE') {
+          sh label: "RUNNING TESTS...", script: "#!/usr/bin/bash \nset -o pipefail;  DEBUG=true python_version='${cell.python}' cython='${cell.cython}' build_dir='${WORKSPACE}/build' cassandra_dtest_dir='${cassandra_dtest_dir}' .build/docker/run-tests.sh ${cell.step} '${cell.split}/${splits}' ${cell.jdk} 2>&1 | tee ${logfile}"
```
Comment (michaelsembwever) [inline .jenkins/Jenkinsfile:266]: note, a script failure != test failures.

SUCCESS : scripted exited 0 and all tests passed
UNSTABLE : script exited 0 but there were test failures
FAILURE : script exited 1 (not related to test failures)


It should be the `junit` publisher that changes the stageResult to UNSTABLE for us.

File: .jenkins/Jenkinsfile (hunk 20 above)
Comment (michaelsembwever) [reply .jenkins/Jenkinsfile:250]: done in my commit.

File: .jenkins/Jenkinsfile (hunk 13 above)
Comment (michaelsembwever) [reply .jenkins/Jenkinsfile:102]: leave for now. will say when ready to address…

File: .jenkins/Jenkinsfile (hunk 18 above)
Comment (michaelsembwever) [reply .jenkins/Jenkinsfile:223]: done in my commit

File: .jenkins/Jenkinsfile (hunk 19 above)
Comment (michaelsembwever) [reply .jenkins/Jenkinsfile:234]: done in my commit

File: .jenkins/Jenkinsfile (hunk 21 above)
Comment (michaelsembwever) [reply .jenkins/Jenkinsfile:275]: done in my commit

File: .jenkins/Jenkinsfile (hunk 22 above)
Comment (michaelsembwever) [reply .jenkins/Jenkinsfile:286]: done in my commit

File: .jenkins/Jenkinsfile (hunk 26 above)
Comment (michaelsembwever) [reply .jenkins/Jenkinsfile:308]: done in my commit

File: .build/docker/_docker_run.sh (hunk 25 above)
Comment (HadesArchitect) [reply .build/docker/_docker_run.sh:34]: We can use preserved stashes between runs, another option will be artifacts. I'll explore possible options.

**Pull request closed without merging**